Bottom: 032824e52fe1ca4c5fb1138c496adbb3d544c9c2
Top:    cf39c825f3196b621175d883975a4ddf740633a6
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-24 01:10:02 +0100

draw.c: Draw pins, pads and vias (for non-gui HIDs) from DrawLayerGroup()


---

diff --git a/src/draw.c b/src/draw.c
index de85b06..b04a362 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -84,7 +84,7 @@ static const BoxType *clip_box = NULL;
 static void Redraw (bool, BoxTypePtr);
 static void DrawEverything (BoxTypePtr);
 static void DrawPPV (int group, const BoxType *);
-static int DrawLayerGroup (int, const BoxType *);
+static void DrawLayerGroup (int, const BoxType *);
 static void DrawPinOrViaLowLevel (PinTypePtr, bool);
 static void DrawPlainPin (PinTypePtr, bool);
 static void DrawPlainVia (PinTypePtr, bool);
@@ -312,11 +312,9 @@ static void
 PrintAssembly (const BoxType * drawn_area, int side_group, int swap_ident)
 {
   int save_swap = SWAP_IDENT;
-  int side = swap_ident ? SOLDER_LAYER : COMPONENT_LAYER;
 
   gui->set_draw_faded (Output.fgGC, 1);
   DrawLayerGroup (side_group, drawn_area);
-  DrawPPV (side_group, drawn_area);
   gui->set_draw_faded (Output.fgGC, 0);
 
   /* draw package */
@@ -381,25 +379,9 @@ DrawEverything (BoxTypePtr drawn_area)
       int group = drawn_groups[i];
 
       if (gui->set_layer (0, group, 0))
-	{
-	  if (DrawLayerGroup (group, drawn_area) && !gui->gui)
-	    {
-	      r_search (PCB->Data->pin_tree, drawn_area, NULL, pin_callback, NULL);
-	      r_search (PCB->Data->via_tree, drawn_area, NULL, via_callback, NULL);
-	      /* draw element pads */
-	      if (group == component || group == solder)
-		{
-                  side = (group == solder) ? SOLDER_LAYER : COMPONENT_LAYER;
-		  r_search (PCB->Data->pad_tree, drawn_area, NULL, pad_callback, &side);
-		}
-
-	      /* draw holes */
-	      plated = -1;
-	      r_search (PCB->Data->pin_tree, drawn_area, NULL, hole_callback, &plated);
-	      r_search (PCB->Data->via_tree, drawn_area, NULL, hole_callback, &plated);
-	    }
-	}
+        DrawLayerGroup (group, drawn_area);
     }
+
   if (TEST_FLAG (CHECKPLANESFLAG, PCB) && gui->gui)
     return;
 
@@ -812,16 +794,18 @@ DrawLayer (LayerTypePtr Layer, const BoxType * screen)
  * draws one layer group.  Returns non-zero if pins and pads should be
  * drawn with this group.
  */
-static int
-DrawLayerGroup (int group, const BoxType * screen)
+static void
+DrawLayerGroup (int group, const BoxType * drawn_area)
 {
   int i, rv = 1;
   int layernum;
+  int plated;
+  int side;
   LayerTypePtr Layer;
   int n_entries = PCB->LayerGroups.Number[group];
   Cardinal *layers = PCB->LayerGroups.Entries[group];
 
-  clip_box = screen;
+  clip_box = drawn_area;
   for (i = n_entries - 1; i >= 0; i--)
     {
       layernum = layers[i];
@@ -830,11 +814,13 @@ DrawLayerGroup (int group, const BoxType * screen)
 	  strcmp (Layer->Name, "route") == 0)
 	rv = 0;
       if (layernum < max_copper_layer && Layer->On)
-	DrawLayerCommon (Layer, screen, true);
+	DrawLayerCommon (Layer, drawn_area, true);
     }
   if (n_entries > 1)
     rv = 1;
-  return rv;
+
+  if (rv == 1 && !gui->gui)
+    DrawPPV (group, drawn_area);
 }
 
 /* ---------------------------------------------------------------------------
